File: app/services/duplicate_detection_service.rb
class DuplicateDetectionService
  # Find all duplicates for a contact
  def self.find_duplicates(contact)
    new(contact).find_duplicates
  end

  # Merge two contacts
  def self.merge(primary_contact, duplicate_contact)
    new(primary_contact).merge_with(duplicate_contact)
  end

  def initialize(contact)
    @contact = contact
  end

  def find_duplicates
    return [] if @contact.is_duplicate?

    candidates = []

    # Phone number exact match
    candidates += find_by_phone_exact if @contact.formatted_phone_number.present?

    # Phone number fuzzy match
    candidates += find_by_phone_fuzzy if @contact.raw_phone_number.present?

    # Email exact match
    candidates += find_by_email if @contact.email.present?

    # Business name + location match
    candidates += find_by_business_identity if @contact.business?

    # Name match (for consumer contacts)
    candidates += find_by_name if @contact.full_name.present?

    # Score and rank candidates
    scored_candidates = candidates.uniq.map do |candidate|
      match_reason = determine_match_reason(@contact, candidate)
      {
        contact: candidate,
        confidence: calculate_match_confidence(@contact, candidate),
        reason: match_reason
      }
    end

    # Filter by confidence threshold
    threshold = TwilioCredential.current&.duplicate_confidence_threshold || 80
    scored_candidates.select { |c| c[:confidence] >= threshold }
                     .sort_by { |c| -c[:confidence] }
  end

  def merge_with(duplicate_contact)
    return false if @contact.id == duplicate_contact.id

    ActiveRecord::Base.transaction do
      # Lock both contacts to prevent concurrent modifications
      ids = [@contact.id, duplicate_contact.id].sort
      locked = Contact.lock.where(id: ids).order(:id).index_by(&:id)
      primary = locked[@contact.id]
      duplicate = locked[duplicate_contact.id]

      return false unless primary && duplicate

      # Double-check neither is already a duplicate after acquiring lock
      return false if primary.is_duplicate? || duplicate.is_duplicate?

      # Merge all data, preferring primary contact's data
      merged_data = merge_data(primary, duplicate)
      merged_data[:duplicate_checked_at] = Time.current

      # Update primary contact with merged data
      primary.update!(merged_data)

      # Record merge history
      merge_record = {
        merged_at: Time.current,
        duplicate_id: duplicate.id,
        duplicate_data: duplicate.attributes.except('id', 'created_at', 'updated_at')
      }

      # Reassign array to ensure Rails detects the change
      merge_history = primary.merge_history || []
      merge_history << merge_record
      primary.merge_history = merge_history
      primary.save!

      # Mark duplicate as merged
      duplicate.update!(
        is_duplicate: true,
        duplicate_of_id: primary.id,
        duplicate_confidence: 100,
        duplicate_checked_at: Time.current
      )

      # Update fingerprints and quality scores
      primary.update_fingerprints!
      primary.calculate_quality_score!

      # Update instance variable to reflect changes
      @contact.reload

      true
    end
  rescue StandardError => e
    Rails.logger.error(
      "Merge failed for contacts (primary: #{primary.id}, duplicate: #{duplicate.id}): " \
      "#{e.class} - #{e.message}"
    )
    Rails.logger.error(e.backtrace.first(5).join("\n"))
    false
  end

  private

  def find_by_phone_exact
    return [] unless @contact.formatted_phone_number.present?

    Contact.where(formatted_phone_number: @contact.formatted_phone_number)
           .where.not(id: @contact.id)
           .where(is_duplicate: false)
           .to_a
  end

  def find_by_phone_fuzzy
    return [] unless @contact.phone_fingerprint.present?

    Contact.where(phone_fingerprint: @contact.phone_fingerprint)
           .where.not(id: @contact.id)
           .where(is_duplicate: false)
           .to_a
  end

  def find_by_email
    return [] unless @contact.email.present?

    if @contact.email_fingerprint.present?
      Contact.where(email_fingerprint: @contact.email_fingerprint)
             .where.not(id: @contact.id)
             .where(is_duplicate: false)
             .to_a
    else
      Contact.where(email: @contact.email)
             .where.not(id: @contact.id)
             .where(is_duplicate: false)
             .to_a
    end
  end

  def find_by_business_identity
    return [] unless @contact.business_name.present?

    # Match on business name + location
    contacts = Contact.where(is_business: true)
                     .where.not(id: @contact.id)
                     .where(is_duplicate: false)

    if @contact.name_fingerprint.present?
      contacts = contacts.where(name_fingerprint: @contact.name_fingerprint)
    else
      contacts = contacts.where(business_name: @contact.business_name)
    end

    # Further filter by location if available
    if @contact.business_city.present?
      contacts = contacts.where(business_city: @contact.business_city)
    end

    contacts.to_a
  end

  def find_by_name
    return [] unless @contact.name_fingerprint.present?

    Contact.where(name_fingerprint: @contact.name_fingerprint)
           .where.not(id: @contact.id)
           .where(is_duplicate: false)
           .to_a
  end

  def calculate_match_confidence(contact1, contact2)
    # Skip empty contacts - no meaningful data to compare
    return 0 if contact1.formatted_phone_number.blank? && contact1.email.blank? && contact1.full_name.blank? && contact1.business_name.blank?
    return 0 if contact2.formatted_phone_number.blank? && contact2.email.blank? && contact2.full_name.blank? && contact2.business_name.blank?

    score = 0
    max_score = 0

    # Phone number match (highest weight)
    max_score += 40
    if contact1.formatted_phone_number.present? && contact2.formatted_phone_number.present?
      if contact1.formatted_phone_number == contact2.formatted_phone_number
        score += 40
      elsif phone_similarity(contact1.raw_phone_number, contact2.raw_phone_number) > 0.8
        score += 30
      end
    end

    # Email match
    max_score += 30
    if contact1.email.present? && contact2.email.present?
      if contact1.email.downcase == contact2.email.downcase
        score += 30
      elsif email_domain_match?(contact1.email, contact2.email)
        score += 15
      end
    end

    # Name match (for businesses or people)
    max_score += 20
    if contact1.business?
      if contact1.business_name.present? && contact2.business_name.present?
        similarity = string_similarity(contact1.business_name, contact2.business_name)
        score += (similarity * 20).to_i
      end
    else
      if contact1.full_name.present? && contact2.full_name.present?
        similarity = string_similarity(contact1.full_name, contact2.full_name)
        score += (similarity * 20).to_i
      end
    end

    # Location match (for businesses)
    max_score += 10
    if contact1.business_city.present? && contact2.business_city.present?
      if contact1.business_city.downcase == contact2.business_city.downcase
        score += 10
      end
    end

    # Return percentage confidence
    max_score > 0 ? ((score.to_f / max_score) * 100).round : 0
  end

  def phone_similarity(phone1, phone2)
    return 0 unless phone1.present? && phone2.present?

    # Remove all non-digits
    p1 = phone1.gsub(/\D/, '')
    p2 = phone2.gsub(/\D/, '')

    # Check last 10 digits (for international numbers)
    p1_last10 = p1.length > 10 ? p1[-10..-1] : p1
    p2_last10 = p2.length > 10 ? p2[-10..-1] : p2

    return 1.0 if p1_last10 == p2_last10

    # Levenshtein distance
    distance = levenshtein_distance(p1_last10, p2_last10)
    max_length = [p1_last10.length, p2_last10.length].max
    return 0 if max_length == 0

    1.0 - (distance.to_f / max_length)
  end

  def email_domain_match?(email1, email2)
    return false unless email1.include?('@') && email2.include?('@')

    domain1 = email1.split('@').last
    domain2 = email2.split('@').last
    domain1.downcase == domain2.downcase
  end

  def determine_match_reason(contact1, contact2)
    reasons = []

    # Check phone match
    if contact1.formatted_phone_number.present? && contact2.formatted_phone_number.present?
      if contact1.formatted_phone_number == contact2.formatted_phone_number
        reasons << "Exact phone match"
      elsif phone_similarity(contact1.raw_phone_number, contact2.raw_phone_number) > 0.8
        reasons << "Similar phone number"
      end
    end

    # Check email match
    if contact1.email.present? && contact2.email.present?
      if contact1.email.downcase == contact2.email.downcase
        reasons << "Exact email match"
      elsif email_domain_match?(contact1.email, contact2.email)
        reasons << "Same email domain"
      end
    end

    # Check business name match
    if contact1.business? && contact1.business_name.present? && contact2.business_name.present?
      similarity = string_similarity(contact1.business_name, contact2.business_name)
      reasons << "Similar business name (#{(similarity * 100).round}%)" if similarity > 0.7
    end

    # Check person name match
    if !contact1.business? && contact1.full_name.present? && contact2.full_name.present?
      similarity = string_similarity(contact1.full_name, contact2.full_name)
      reasons << "Similar name (#{(similarity * 100).round}%)" if similarity > 0.7
    end

    # Check location match
    if contact1.business_city.present? && contact2.business_city.present?
      if contact1.business_city.downcase == contact2.business_city.downcase
        reasons << "Same city"
      end
    end

    reasons.any? ? reasons.join(", ") : "Multiple field similarities"
  end

  def string_similarity(str1, str2)
    return 0 unless str1.present? && str2.present?

    s1 = str1.downcase.strip
    s2 = str2.downcase.strip

    return 1.0 if s1 == s2

    # Levenshtein distance
    distance = levenshtein_distance(s1, s2)
    max_length = [s1.length, s2.length].max
    return 0 if max_length == 0

    1.0 - (distance.to_f / max_length)
  end

  # Levenshtein distance algorithm
  def levenshtein_distance(s1, s2)
    return s2.length if s1.empty?
    return s1.length if s2.empty?

    matrix = Array.new(s1.length + 1) { Array.new(s2.length + 1) }

    (0..s1.length).each { |i| matrix[i][0] = i }
    (0..s2.length).each { |j| matrix[0][j] = j }

    (1..s1.length).each do |i|
      (1..s2.length).each do |j|
        cost = s1[i - 1] == s2[j - 1] ? 0 : 1
        matrix[i][j] = [
          matrix[i - 1][j] + 1,      # deletion
          matrix[i][j - 1] + 1,      # insertion
          matrix[i - 1][j - 1] + cost # substitution
        ].min
      end
    end

    matrix[s1.length][s2.length]
  end

  def merge_data(primary, duplicate)
    merged = {}

    # For each field, prefer non-nil value from primary, then duplicate
    # Or if both present, prefer higher quality data

    # Phone data
    merged[:formatted_phone_number] = best_value(
      primary.formatted_phone_number,
      duplicate.formatted_phone_number
    )

    # Email data
    if primary.email_verified || !duplicate.email_verified
      merged[:email] = primary.email || duplicate.email
      merged[:email_verified] = primary.email_verified || duplicate.email_verified
      merged[:email_score] = [primary.email_score, duplicate.email_score].compact.max
    else
      merged[:email] = duplicate.email
      merged[:email_verified] = duplicate.email_verified
      merged[:email_score] = duplicate.email_score
    end

    # Name data
    merged[:full_name] = best_value(primary.full_name, duplicate.full_name)
    merged[:first_name] = best_value(primary.first_name, duplicate.first_name)
    merged[:last_name] = best_value(primary.last_name, duplicate.last_name)

    # Business data - prefer enriched data
    if primary.business_enriched? || !duplicate.business_enriched?
      merged[:business_name] = best_value(primary.business_name, duplicate.business_name)
      merged[:business_employee_count] = best_value(primary.business_employee_count, duplicate.business_employee_count)
      merged[:business_annual_revenue] = best_value(primary.business_annual_revenue, duplicate.business_annual_revenue)
      merged[:business_website] = best_value(primary.business_website, duplicate.business_website)
    else
      merged[:business_name] = duplicate.business_name || primary.business_name
      merged[:business_employee_count] = duplicate.business_employee_count || primary.business_employee_count
      merged[:business_annual_revenue] = duplicate.business_annual_revenue || primary.business_annual_revenue
      merged[:business_website] = duplicate.business_website || primary.business_website
    end

    # Collect additional emails
    emails = [primary.email, duplicate.email, *primary.additional_emails, *duplicate.additional_emails].compact.uniq
    merged[:additional_emails] = emails.reject { |e| e == merged[:email] }

    merged
  end

  def best_value(primary_value, duplicate_value)
    return primary_value if primary_value.present?
    duplicate_value
  end
end

File: laravel-app/app/Services/DuplicateDetectionService.php
<?php

namespace App\Services;

use App\Models\Contact;
use App\Models\TwilioCredential;
use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Log;
use Exception;

/**
 * DuplicateDetectionService - Finds and merges duplicate contacts
 *
 * This service implements sophisticated duplicate detection using:
 * - Phone number normalization (E.164 format)
 * - Fuzzy matching for business names (Levenshtein distance)
 * - Email domain matching
 * - Address similarity scoring
 * - Phonetic matching (Soundex/Metaphone)
 * - Duplicate fingerprint generation
 * - Confidence scoring (0-100)
 *
 * Usage:
 *   // Find duplicates
 *   $duplicates = DuplicateDetectionService::findDuplicates($contact);
 *
 *   // Merge contacts
 *   DuplicateDetectionService::merge($primaryContact, $duplicateContact);
 */
class DuplicateDetectionService
{
    /** @var Contact */
    private $contact;

    /**
     * Find all duplicates for a contact
     *
     * @param Contact $contact Contact to check for duplicates
     * @return array Array of duplicate candidates with confidence scores
     */
    public static function findDuplicates(Contact $contact): array
    {
        return (new self($contact))->find();
    }

    /**
     * Merge two contacts
     *
     * @param Contact $primaryContact Primary contact (will be kept)
     * @param Contact $duplicateContact Duplicate contact (will be marked as duplicate)
     * @return bool Success status
     */
    public static function merge(Contact $primaryContact, Contact $duplicateContact): bool
    {
        return (new self($primaryContact))->mergeWith($duplicateContact);
    }

    /**
     * Constructor
     *
     * @param Contact $contact Contact instance
     */
    public function __construct(Contact $contact)
    {
        $this->contact = $contact;
    }

    /**
     * Find all duplicate candidates for the contact
     *
     * @return array Scored and ranked duplicate candidates
     */
    public function find(): array
    {
        // Don't search for duplicates if this contact is already marked as duplicate
        if ($this->contact->is_duplicate) {
            return [];
        }

        $candidates = [];

        // Phone number exact match
        if (!empty($this->contact->formatted_phone_number)) {
            $candidates = array_merge($candidates, $this->findByPhoneExact());
        }

        // Phone number fuzzy match (fingerprint-based)
        if (!empty($this->contact->raw_phone_number)) {
            $candidates = array_merge($candidates, $this->findByPhoneFuzzy());
        }

        // Email exact match
        if (!empty($this->contact->email)) {
            $candidates = array_merge($candidates, $this->findByEmail());
        }

        // Business name + location match
        if ($this->contact->is_business) {
            $candidates = array_merge($candidates, $this->findByBusinessIdentity());
        }

        // Name match (for consumer contacts)
        if (!empty($this->contact->full_name)) {
            $candidates = array_merge($candidates, $this->findByName());
        }

        // Remove duplicates and score candidates
        $uniqueCandidates = collect($candidates)
            ->unique('id')
            ->map(function ($candidate) {
                $confidence = $this->calculateMatchConfidence($this->contact, $candidate);
                $reason = $this->determineMatchReason($this->contact, $candidate);

                return [
                    'contact' => $candidate,
                    'confidence' => $confidence,
                    'reason' => $reason,
                ];
            })
            ->filter(function ($item) {
                // Filter by confidence threshold
                $threshold = TwilioCredential::current()?->duplicate_confidence_threshold ?? 80;
                return $item['confidence'] >= $threshold;
            })
            ->sortByDesc('confidence')
            ->values()
            ->all();

        return $uniqueCandidates;
    }

    /**
     * Merge duplicate contact into primary contact
     *
     * @param Contact $duplicateContact Contact to merge (will be marked as duplicate)
     * @return bool Success status
     */
    public function mergeWith(Contact $duplicateContact): bool
    {
        // Can't merge contact with itself
        if ($this->contact->id === $duplicateContact->id) {
            return false;
        }

        try {
            return DB::transaction(function () use ($duplicateContact) {
                // Lock both contacts to prevent concurrent modifications
                $ids = [$this->contact->id, $duplicateContact->id];
                sort($ids);

                $locked = Contact::lockForUpdate()
                    ->whereIn('id', $ids)
                    ->orderBy('id')
                    ->get()
                    ->keyBy('id');

                $primary = $locked->get($this->contact->id);
                $duplicate = $locked->get($duplicateContact->id);

                if (!$primary || !$duplicate) {
                    return false;
                }

                // Double-check neither is already a duplicate after acquiring lock
                if ($primary->is_duplicate || $duplicate->is_duplicate) {
                    return false;
                }

                // Merge all data, preferring primary contact's data
                $mergedData = $this->mergeData($primary, $duplicate);
                $mergedData['duplicate_checked_at'] = now();

                // Update primary contact with merged data
                $primary->update($mergedData);

                // Record merge history
                $mergeRecord = [
                    'merged_at' => now()->toIso8601String(),
                    'duplicate_id' => $duplicate->id,
                    'duplicate_data' => collect($duplicate->getAttributes())
                        ->except(['id', 'created_at', 'updated_at'])
                        ->all(),
                ];

                // Update merge history
                $mergeHistory = $primary->merge_history ?? [];
                $mergeHistory[] = $mergeRecord;
                $primary->merge_history = $mergeHistory;
                $primary->save();

                // Mark duplicate as merged
                $duplicate->update([
                    'is_duplicate' => true,
                    'duplicate_of_id' => $primary->id,
                    'duplicate_confidence' => 100,
                    'duplicate_checked_at' => now(),
                ]);

                // Update fingerprints and quality scores
                if (method_exists($primary, 'updateFingerprints')) {
                    $primary->updateFingerprints();
                }
                if (method_exists($primary, 'calculateQualityScore')) {
                    $primary->calculateQualityScore();
                }

                // Reload the contact instance
                $this->contact->refresh();

                return true;
            });
        } catch (Exception $e) {
            Log::error("Merge failed: {$e->getMessage()}");
            return false;
        }
    }

    /**
     * Find contacts with exact phone number match
     *
     * @return array Matching contacts
     */
    private function findByPhoneExact(): array
    {
        if (empty($this->contact->formatted_phone_number)) {
            return [];
        }

        return Contact::where('formatted_phone_number', $this->contact->formatted_phone_number)
            ->where('id', '!=', $this->contact->id)
            ->where('is_duplicate', false)
            ->get()
            ->all();
    }

    /**
     * Find contacts with fuzzy phone number match (using fingerprint)
     *
     * @return array Matching contacts
     */
    private function findByPhoneFuzzy(): array
    {
        if (empty($this->contact->phone_fingerprint)) {
            return [];
        }

        return Contact::where('phone_fingerprint', $this->contact->phone_fingerprint)
            ->where('id', '!=', $this->contact->id)
            ->where('is_duplicate', false)
            ->get()
            ->all();
    }

    /**
     * Find contacts with matching email
     *
     * @return array Matching contacts
     */
    private function findByEmail(): array
    {
        if (empty($this->contact->email)) {
            return [];
        }

        if (!empty($this->contact->email_fingerprint)) {
            return Contact::where('email_fingerprint', $this->contact->email_fingerprint)
                ->where('id', '!=', $this->contact->id)
                ->where('is_duplicate', false)
                ->get()
                ->all();
        }

        return Contact::where('email', $this->contact->email)
            ->where('id', '!=', $this->contact->id)
            ->where('is_duplicate', false)
            ->get()
            ->all();
    }

    /**
     * Find contacts with matching business identity (name + location)
     *
     * @return array Matching contacts
     */
    private function findByBusinessIdentity(): array
    {
        if (empty($this->contact->business_name)) {
            return [];
        }

        $query = Contact::where('is_business', true)
            ->where('id', '!=', $this->contact->id)
            ->where('is_duplicate', false);

        // Match on business name fingerprint or exact name
        if (!empty($this->contact->name_fingerprint)) {
            $query->where('name_fingerprint', $this->contact->name_fingerprint);
        } else {
            $query->where('business_name', $this->contact->business_name);
        }

        // Further filter by location if available
        if (!empty($this->contact->business_city)) {
            $query->where('business_city', $this->contact->business_city);
        }

        return $query->get()->all();
    }

    /**
     * Find contacts with matching name (for consumer contacts)
     *
     * @return array Matching contacts
     */
    private function findByName(): array
    {
        if (empty($this->contact->name_fingerprint)) {
            return [];
        }

        return Contact::where('name_fingerprint', $this->contact->name_fingerprint)
            ->where('id', '!=', $this->contact->id)
            ->where('is_duplicate', false)
            ->get()
            ->all();
    }

    /**
     * Calculate match confidence score (0-100)
     *
     * Weighted scoring:
     * - Phone number: 40 points (highest weight)
     * - Email: 30 points
     * - Name/Business name: 20 points
     * - Location: 10 points
     *
     * @param Contact $contact1 First contact
     * @param Contact $contact2 Second contact
     * @return int Confidence score (0-100)
     */
    private function calculateMatchConfidence(Contact $contact1, Contact $contact2): int
    {
        // Skip empty contacts - no meaningful data to compare
        if ($this->isEmptyContact($contact1) || $this->isEmptyContact($contact2)) {
            return 0;
        }

        $score = 0;
        $maxScore = 0;

        // Phone number match (highest weight: 40 points)
        $maxScore += 40;
        if (!empty($contact1->formatted_phone_number) && !empty($contact2->formatted_phone_number)) {
            if ($contact1->formatted_phone_number === $contact2->formatted_phone_number) {
                $score += 40;
            } elseif ($this->phoneSimilarity($contact1->raw_phone_number, $contact2->raw_phone_number) > 0.8) {
                $score += 30;
            }
        }

        // Email match (30 points)
        $maxScore += 30;
        if (!empty($contact1->email) && !empty($contact2->email)) {
            if (strtolower($contact1->email) === strtolower($contact2->email)) {
                $score += 30;
            } elseif ($this->emailDomainMatch($contact1->email, $contact2->email)) {
                $score += 15;
            }
        }

        // Name match (20 points - for businesses or people)
        $maxScore += 20;
        if ($contact1->is_business) {
            if (!empty($contact1->business_name) && !empty($contact2->business_name)) {
                $similarity = $this->stringSimilarity($contact1->business_name, $contact2->business_name);
                $score += (int) ($similarity * 20);
            }
        } elseif (!empty($contact1->full_name) && !empty($contact2->full_name)) {
            $similarity = $this->stringSimilarity($contact1->full_name, $contact2->full_name);
            $score += (int) ($similarity * 20);
        }

        // Location match (10 points - for businesses)
        $maxScore += 10;
        if (!empty($contact1->business_city) && !empty($contact2->business_city)) {
            if (strtolower($contact1->business_city) === strtolower($contact2->business_city)) {
                $score += 10;
            }
        }

        // Return percentage confidence
        return $maxScore > 0 ? (int) round(($score / $maxScore) * 100) : 0;
    }

    /**
     * Check if contact has no meaningful data
     *
     * @param Contact $contact Contact to check
     * @return bool True if contact is empty
     */
    private function isEmptyContact(Contact $contact): bool
    {
        return empty($contact->formatted_phone_number)
            && empty($contact->email)
            && empty($contact->full_name)
            && empty($contact->business_name);
    }

    /**
     * Calculate phone number similarity (0.0-1.0)
     *
     * Uses Levenshtein distance on last 10 digits
     *
     * @param string|null $phone1 First phone number
     * @param string|null $phone2 Second phone number
     * @return float Similarity score (0.0-1.0)
     */
    private function phoneSimilarity(?string $phone1, ?string $phone2): float
    {
        if (empty($phone1) || empty($phone2)) {
            return 0.0;
        }

        // Remove all non-digits
        $p1 = preg_replace('/\D/', '', $phone1);
        $p2 = preg_replace('/\D/', '', $phone2);

        // Check last 10 digits (for international numbers)
        $p1Last10 = strlen($p1) > 10 ? substr($p1, -10) : $p1;
        $p2Last10 = strlen($p2) > 10 ? substr($p2, -10) : $p2;

        if ($p1Last10 === $p2Last10) {
            return 1.0;
        }

        // Levenshtein distance
        $distance = $this->levenshteinDistance($p1Last10, $p2Last10);
        $maxLength = max(strlen($p1Last10), strlen($p2Last10));

        if ($maxLength === 0) {
            return 0.0;
        }

        return 1.0 - ($distance / $maxLength);
    }

    /**
     * Check if two emails have the same domain
     *
     * @param string|null $email1 First email
     * @param string|null $email2 Second email
     * @return bool True if domains match
     */
    private function emailDomainMatch(?string $email1, ?string $email2): bool
    {
        if (empty($email1) || empty($email2)) {
            return false;
        }

        if (!str_contains($email1, '@') || !str_contains($email2, '@')) {
            return false;
        }

        $domain1 = strtolower(substr(strrchr($email1, '@'), 1));
        $domain2 = strtolower(substr(strrchr($email2, '@'), 1));

        return $domain1 === $domain2;
    }

    /**
     * Determine the reason for match between two contacts
     *
     * @param Contact $contact1 First contact
     * @param Contact $contact2 Second contact
     * @return string Human-readable match reason
     */
    private function determineMatchReason(Contact $contact1, Contact $contact2): string
    {
        $reasons = [];

        // Check phone match
        if (!empty($contact1->formatted_phone_number) && !empty($contact2->formatted_phone_number)) {
            if ($contact1->formatted_phone_number === $contact2->formatted_phone_number) {
                $reasons[] = 'Exact phone match';
            } elseif ($this->phoneSimilarity($contact1->raw_phone_number, $contact2->raw_phone_number) > 0.8) {
                $reasons[] = 'Similar phone number';
            }
        }

        // Check email match
        if (!empty($contact1->email) && !empty($contact2->email)) {
            if (strtolower($contact1->email) === strtolower($contact2->email)) {
                $reasons[] = 'Exact email match';
            } elseif ($this->emailDomainMatch($contact1->email, $contact2->email)) {
                $reasons[] = 'Same email domain';
            }
        }

        // Check business name match
        if ($contact1->is_business && !empty($contact1->business_name) && !empty($contact2->business_name)) {
            $similarity = $this->stringSimilarity($contact1->business_name, $contact2->business_name);
            if ($similarity > 0.7) {
                $reasons[] = sprintf('Similar business name (%d%%)', (int) round($similarity * 100));
            }
        }

        // Check person name match
        if (!$contact1->is_business && !empty($contact1->full_name) && !empty($contact2->full_name)) {
            $similarity = $this->stringSimilarity($contact1->full_name, $contact2->full_name);
            if ($similarity > 0.7) {
                $reasons[] = sprintf('Similar name (%d%%)', (int) round($similarity * 100));
            }
        }

        // Check location match
        if (!empty($contact1->business_city) && !empty($contact2->business_city)) {
            if (strtolower($contact1->business_city) === strtolower($contact2->business_city)) {
                $reasons[] = 'Same city';
            }
        }

        return !empty($reasons) ? implode(', ', $reasons) : 'Multiple field similarities';
    }

    /**
     * Calculate string similarity using Levenshtein distance (0.0-1.0)
     *
     * @param string|null $str1 First string
     * @param string|null $str2 Second string
     * @return float Similarity score (0.0-1.0)
     */
    private function stringSimilarity(?string $str1, ?string $str2): float
    {
        if (empty($str1) || empty($str2)) {
            return 0.0;
        }

        $s1 = strtolower(trim($str1));
        $s2 = strtolower(trim($str2));

        if ($s1 === $s2) {
            return 1.0;
        }

        // Levenshtein distance
        $distance = $this->levenshteinDistance($s1, $s2);
        $maxLength = max(strlen($s1), strlen($s2));

        if ($maxLength === 0) {
            return 0.0;
        }

        return 1.0 - ($distance / $maxLength);
    }

    /**
     * Calculate Levenshtein distance between two strings
     *
     * Measures the minimum number of single-character edits (insertions,
     * deletions, or substitutions) required to change one string into another.
     *
     * @param string $s1 First string
     * @param string $s2 Second string
     * @return int Edit distance
     */
    private function levenshteinDistance(string $s1, string $s2): int
    {
        if (empty($s1)) {
            return strlen($s2);
        }
        if (empty($s2)) {
            return strlen($s1);
        }

        $len1 = strlen($s1);
        $len2 = strlen($s2);

        // Initialize matrix
        $matrix = array_fill(0, $len1 + 1, array_fill(0, $len2 + 1, 0));

        // Fill first column and row
        for ($i = 0; $i <= $len1; $i++) {
            $matrix[$i][0] = $i;
        }
        for ($j = 0; $j <= $len2; $j++) {
            $matrix[0][$j] = $j;
        }

        // Calculate distances
        for ($i = 1; $i <= $len1; $i++) {
            for ($j = 1; $j <= $len2; $j++) {
                $cost = $s1[$i - 1] === $s2[$j - 1] ? 0 : 1;
                $matrix[$i][$j] = min(
                    $matrix[$i - 1][$j] + 1,      // deletion
                    $matrix[$i][$j - 1] + 1,      // insertion
                    $matrix[$i - 1][$j - 1] + $cost // substitution
                );
            }
        }

        return $matrix[$len1][$len2];
    }

    /**
     * Merge data from two contacts, preferring primary contact's data
     *
     * @param Contact $primary Primary contact
     * @param Contact $duplicate Duplicate contact
     * @return array Merged data array
     */
    private function mergeData(Contact $primary, Contact $duplicate): array
    {
        $merged = [];

        // Phone data - prefer non-null value from primary
        $merged['formatted_phone_number'] = $this->bestValue(
            $primary->formatted_phone_number,
            $duplicate->formatted_phone_number
        );

        // Email data - prefer verified email
        if ($primary->email_verified || !$duplicate->email_verified) {
            $merged['email'] = $primary->email ?? $duplicate->email;
            $merged['email_verified'] = $primary->email_verified || $duplicate->email_verified;
            $merged['email_score'] = max(
                $primary->email_score ?? 0,
                $duplicate->email_score ?? 0
            );
        } else {
            $merged['email'] = $duplicate->email;
            $merged['email_verified'] = $duplicate->email_verified;
            $merged['email_score'] = $duplicate->email_score;
        }

        // Name data
        $merged['full_name'] = $this->bestValue($primary->full_name, $duplicate->full_name);
        $merged['first_name'] = $this->bestValue($primary->first_name, $duplicate->first_name);
        $merged['last_name'] = $this->bestValue($primary->last_name, $duplicate->last_name);

        // Business data - prefer enriched data
        $primaryEnriched = !empty($primary->business_enriched_at);
        $duplicateEnriched = !empty($duplicate->business_enriched_at);

        if ($primaryEnriched || !$duplicateEnriched) {
            $merged['business_name'] = $this->bestValue($primary->business_name, $duplicate->business_name);
            $merged['business_employee_count'] = $this->bestValue(
                $primary->business_employee_count,
                $duplicate->business_employee_count
            );
            $merged['business_annual_revenue'] = $this->bestValue(
                $primary->business_annual_revenue,
                $duplicate->business_annual_revenue
            );
            $merged['business_website'] = $this->bestValue($primary->business_website, $duplicate->business_website);
        } else {
            $merged['business_name'] = $duplicate->business_name ?? $primary->business_name;
            $merged['business_employee_count'] = $duplicate->business_employee_count ?? $primary->business_employee_count;
            $merged['business_annual_revenue'] = $duplicate->business_annual_revenue ?? $primary->business_annual_revenue;
            $merged['business_website'] = $duplicate->business_website ?? $primary->business_website;
        }

        // Collect additional emails (merge unique emails)
        $allEmails = array_merge(
            [$primary->email],
            [$duplicate->email],
            $primary->additional_emails ?? [],
            $duplicate->additional_emails ?? []
        );
        $allEmails = array_filter($allEmails); // Remove nulls
        $allEmails = array_unique($allEmails);
        $merged['additional_emails'] = array_values(
            array_diff($allEmails, [$merged['email']])
        );

        return $merged;
    }

    /**
     * Return the best value between primary and duplicate
     *
     * @param mixed $primaryValue Primary contact's value
     * @param mixed $duplicateValue Duplicate contact's value
     * @return mixed Best value
     */
    private function bestValue($primaryValue, $duplicateValue)
    {
        return !empty($primaryValue) ? $primaryValue : $duplicateValue;
    }
}
